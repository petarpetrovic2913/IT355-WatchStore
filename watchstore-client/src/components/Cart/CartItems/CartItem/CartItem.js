import React, { Component } from 'react';

import { connect } from 'react-redux';
import { removeCartItem, getCart } from '../../../../store/actions/cartActions';
import { Button, Table } from 'semantic-ui-react';

export class CartItem extends Component {
  onDelete = async () => {
    const id = this.props.content.cartItemId;
    await this.props.removeCartItem(id);
    await this.props.getCart();
  };

  render() {
    const { productName } = this.props.content.product;
    const { quantity, totalPrice } = this.props.content;
    return (
      <Table.Row>
        <Table.Cell width={3}>{productName}</Table.Cell>
        <Table.Cell width={3}>{quantity}</Table.Cell>
        <Table.Cell width={3}>{totalPrice}</Table.Cell>
        <Table.Cell width={2}>
          <Button
            color="red"
            onClick={this.onDelete}
            data-testid="delete-button"
          >
            Remove from cart
          </Button>
        </Table.Cell>
      </Table.Row>
    );
  }
}

const mapDispatchToProps = { removeCartItem, getCart };

export default connect(null, mapDispatchToProps)(CartItem);
