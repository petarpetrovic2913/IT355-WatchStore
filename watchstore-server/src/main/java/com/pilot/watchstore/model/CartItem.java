package com.pilot.watchstore.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import java.io.Serializable;
import java.util.Objects;

@Entity
public class CartItem implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long cartItemId;

    @ManyToOne()
    @JoinColumn(name = "cartId")
    @JsonIgnore
    private Cart cart;

    @ManyToOne
    @JoinColumn(name = "productId")
    private Product product;

    private int quantity;
    private double totalPrice;

    public CartItem(){}

    public CartItem(Cart cart, Product product, int quantity, double totalPrice) {
        this.cart = cart;
        this.product = product;
        this.quantity = quantity;
        this.totalPrice = totalPrice;
    }

    public Long getCartItemId() {
        return cartItemId;
    }

    public void setCartItemId(Long cartItemId) {
        this.cartItemId = cartItemId;
    }

    public Cart getCart() {
        return cart;
    }

    public void setCart(Cart cart) {
        this.cart = cart;
    }

    public Product getProduct() {
        return product;
    }

    public void setProduct(Product product) {
        this.product = product;
    }

    public int getQuantity() {
        return quantity;
    }

    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }

    public double getTotalPrice() {
        return totalPrice;
    }

    public void setTotalPrice(double totalPrice) {
        this.totalPrice = totalPrice;
    }

}
