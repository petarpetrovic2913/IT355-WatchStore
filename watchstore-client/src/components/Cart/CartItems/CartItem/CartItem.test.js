import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { CartItem } from './CartItem';

describe('Test CartItem component', () => {
  it('renders without crashing', () => {
    const removeCartItem = jest.fn();
    const getCart = jest.fn();
    const content = {
      product: {
        productName: 'test name'
      },
      totalPrice: 125,
      quantity: 5
    };

    const { container } = render(
      <CartItem
        removeCartItem={removeCartItem}
        getCart={getCart}
        content={content}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should call removeCartItem and getCart on delete', async () => {
    const removeCartItem = jest.fn();
    const getCart = jest.fn();
    const content = {
      cartItemId: 6,
      product: {
        productName: 'test name'
      },
      totalPrice: 125,
      quantity: 5
    };

    const { getByTestId } = render(
      <CartItem
        removeCartItem={removeCartItem}
        getCart={getCart}
        content={content}
      />
    );
    const deleteButton = getByTestId('delete-button');
    fireEvent.click(deleteButton);

    expect(removeCartItem).toHaveBeenCalledWith(6);
    await wait(() => expect(getCart).toHaveBeenCalled());
  });
});
