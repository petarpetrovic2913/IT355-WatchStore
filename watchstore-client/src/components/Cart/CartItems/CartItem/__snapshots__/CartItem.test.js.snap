// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CartItem component renders without crashing 1`] = `
<div>
  <tr
    class=""
  >
    <td
      class="three wide"
    >
      test name
    </td>
    <td
      class="three wide"
    >
      5
    </td>
    <td
      class="three wide"
    >
      125
    </td>
    <td
      class="two wide"
    >
      <button
        class="ui red button"
        data-testid="delete-button"
      >
        Remove from cart
      </button>
    </td>
  </tr>
</div>
`;
